xc7/utils/prjxray_create_equiv_tiles.py

```diff
@@ -0,0 +1,1253 @@
+    # Generate the pb_type.xml file                                          #
+    ##########################################################################
+
+    pb_type_xml = start_pb_type(
+        pb_type,
+        pin_assignments,
+        ['{}_{}'.format(site, pin) for site, pin in all_pb_type_output_pins],
+        ['{}_{}'.format(site, pin) for site, pin in all_pb_type_input_pins],
+        root_pb_type=False
+    )
+
+    interconnect_xml = ET.Element('interconnect')
+
+    site_pbtype = site_directory + "/{0}/{1}.pb_type.xml"
+
+    site_prefixes = get_tile_prefixes(
+        conn, tile_type_pkeys, site_type_pkeys, site_remaps
+    )
+
+    cell_names = {}
+    site_type_ports = {}
+
+    for site_type_pkey in site_type_pkeys:
+        cur.execute(
+            "SELECT name FROM site_type WHERE pkey = ?", (site_type_pkey, )
+        )
+        site_type = cur.fetchone()[0]
+        site_prefix = site_prefixes[site_type_pkey]
+        site_type_path = site_pbtype.format(
+            site_type.lower(), site_type.lower()
+        )
+
+        cell_pb_type = ET.ElementTree()
+        root_element = cell_pb_type.parse(site_type_path)
+        cell_names[site_type] = root_element.attrib['name']
+
+        ports = {}
+        for inputs in root_element.iter('input'):
+            ports[inputs.attrib['name']] = int(inputs.attrib['num_pins'])
+
+        for clocks in root_element.iter('clock'):
+            ports[clocks.attrib['name']] = int(clocks.attrib['num_pins'])
+
+        for outputs in root_element.iter('output'):
+            ports[outputs.attrib['name']] = int(outputs.attrib['num_pins'])
+
+        assert site_type not in site_type_ports, (
+            site_type, site_type_ports.keys()
+        )
+        site_type_ports[site_type] = ports
+
+        attrib = dict(root_element.attrib)
+        include_xml = ET.SubElement(pb_type_xml, 'pb_type', attrib)
+        ET.SubElement(
+            include_xml, XI_INCLUDE, {
+                'href':
+                    site_type_path,
+                'xpointer':
+                    "xpointer(pb_type/child::node()[local-name()!='metadata'])",
+            }
+        )
+
+        metadata_xml = ET.SubElement(include_xml, 'metadata')
+        ET.SubElement(
+            metadata_xml, 'meta', {
+                'name': 'fasm_prefix',
+            }
+        ).text = site_prefix
+
+        # Import pb_type metadata if it exists.
+        if any(child.tag == 'metadata' for child in root_element):
+            ET.SubElement(
+                metadata_xml, XI_INCLUDE, {
+                    'href': site_type_path,
+                    'xpointer': "xpointer(pb_type/metadata/child::node())",
+                }
+            )
+
+    for site_type, site_pin_name in top_level_pin_external:
+        port = find_port(site_pin_name, site_type_ports[site_type])
+        if port is None:
+            print(
+                "*** WARNING *** Didn't find port for name {} for site type {}"
```

Comment:
Should this be a warning or an error?

Reply:
Warning.  Some ports are never connected.